[LLD][ELF][AArch64] Add BTI Aware long branch thunks

smithp35 opened the pull request:
When Branch Target Identification BTI is enabled all indirect branches must target a BTI instruction. A long branch thunk is a source of indirect branches. To date LLD has been assuming that the object producer is responsible for putting a BTI instruction at all places the linker might generate an indirect branch to. This is true for clang, but not for GCC. GCC will elide the BTI instruction when it can prove that there are no indirect branches from outside the translation unit(s). GNU ld was fixed to generate a landing pad stub (gnu ld speak for thunk) for the destination when a long range stub was needed [1].

This means that using GCC compiled objects with LLD may lead to LLD generating an indirect branch to a location without a BTI. The ABI [2] has also been clarified to say that it is a static linker's responsibility to generate a landing pad when the target does not have a BTI.

This patch implements the same mechansim as GNU ld. When the output ELF file is setting the
GNU_PROPERTY_AARCH64_FEATURE_1_BTI property, then we check the destination to see if it has a BTI instruction. If it does not we generate a landing pad consisting of:
BTI c
B <destination>

The B <destination> can be elided if the thunk can be placed so that control flow drops through. For example:
BTI c
<destination>:
This will be common when -ffunction-sections is used.

The landing pad thunks are effectively alternative entry points for the function. Direct branches are unaffected but any linker generated indirect branch needs to use the alternative. We place these as close as possible to the destination section.

There is some further optimization possible. Consider the case:
.text
fn1
...
fn2
...

If we need landing pad thunks for both fn1 and fn2 we could order them so that the thunk for fn1 immediately precedes fn1. This could save a single branch. However I didn't think that would be worth the additional complexity.

[1] https://gcc.gnu.org/bugzilla/show_bug.cgi?id=106671
[2] https://github.com/ARM-software/abi-aa/issues/196

smithp35 committed:
Add BTI Aware long branch thunks

When Branch Target Identification BTI is enabled all indirect
branches must target a BTI instruction. A long branch thunk
is a source of indirect branches. To date LLD has been
assuming that the object producer is responsible for putting
a BTI instruction at all places the linker might generate an
indirect branch to. This is true for clang, but not for GCC.
GCC will elide the BTI instruction when it can prove that
there are no indirect branches from outside the translation
unit(s). GNU ld was fixed to generate a landing pad stub
(gnu ld speak for thunk) for the destination when a long
range stub was needed [1].

This means that using GCC compiled objects with LLD may
lead to LLD generating an indirect branch to a location
without a BTI. The ABI [2] has also been clarified to say
that it is a static linker's responsibility to generate
a landing pad when the target does not have a BTI.

This patch implements the same mechansim as GNU ld. When
the output ELF file is setting the
GNU_PROPERTY_AARCH64_FEATURE_1_BTI property, then we check
the destination to see if it has a BTI instruction. If it
does not we generate a landing pad consisting of:
BTI c
B <destination>

The B <destination> can be elided if the thunk can be placed
so that control flow drops through. For example:
BTI c
<destination>:
This will be common when -ffunction-sections is used.

The landing pad thunks are effectively alternative entry
points for the function. Direct branches are unaffected
but any linker generated indirect branch needs to use
the alternative. We place these as close as possible
to the destination section.

There is some further optimization possible. Consider the
case:
.text
fn1
...
fn2
...

If we need landing pad thunks for both fn1 and fn2 we could
order them so that the thunk for fn1 immediately precedes fn1.
This could save a single branch. However I didn't think that
would be worth the additional complexity.

[1] https://gcc.gnu.org/bugzilla/show_bug.cgi?id=106671
[2] https://github.com/ARM-software/abi-aa/issues/196

smithp35 committed:
Make test more robust to pathname changes

The .dynstr is in between the .plt and the .got.plt. If this
contains a DT_NEEDED tag then the pathname of the shared object
will be used. This path is machine dependent. Don't hard code
the offsets in the .plt in this case. We are only disassembling
the .plt to show that there is a BTI at the start of it.

MaskRay commented:
Thanks for the patch. I am back from vacation and will check this shortly.

FYI: I plan to replace `config->x` with `ctx.arg.x` to move toward the goal to eliminate the global states. It should require minimal changes to this patch.

MaskRay commented:
> Hmm, the tests are passing on my local linux machine, not sure what has happened with Windows. Will need to try and find a Windows machine and investigate.
> 
> [UPDATE] It looks like it is due to the positioning of the .dynstr section, the test's .dynamic section has a DT_NEEDED which has a file path in it, this will affect the offset between the .plt and .got.plt. I'll update the test to be more robust.

If an output DSO is used to link another component, we do `ld.lld -shared -soname=t.so ... -o %t.so` to ensure that the DT_NEEDED entry needs a fixed-length .dynstr entry.

This is error-prone, which is one of the reasons I prefer `rm -rf %t && split-file %s %t && cd %t` :)

MaskRay commented on lld/test/ELF/aarch64-thunk-bti.s:
```diff
@@ -0,0 +1,513 @@
+// RUN: split-file %s %t
+// RUN: llvm-mc -filetype=obj -triple=aarch64 %t/asm -o %t.o
+// RUN: ld.lld --shared --script=%t/lds %t.o -o %t.so --defsym absolute=0xf0000000
+// RUN: llvm-objdump -d --no-show-raw-insn --print-imm-hex --start-address=0x10000000 --stop-address=0x10000028 %t.so | FileCheck --check-prefix=CHECK-SO %s
```
#109553 might be useful to allow removing some llvm-objdump commands, making maintenance more convenient.

`llvm/tools/llvm-objdump/llvm-objdump.cpp:countSkippableZeroBytes` needs to spend time to skip zeroes, so the performance is not as good as multiple llvm-objdump invocations.

smithp35 replied:
I've rewritten the test using OutputSections to need much fewer zeroes.

MaskRay commented on lld/test/ELF/aarch64-thunk-bti.s:
```diff
@@ -0,0 +1,513 @@
+  high PT_LOAD FLAGS(0x1 | 0x4);
+}
+SECTIONS {
+    .text_low 0x10000000 : {
```
Unintended 4-column when `PHDRS` uses 2-column indentation?

smithp35 replied:
ACK. I've made the indentation consistent.

MaskRay commented on lld/test/ELF/aarch64-thunk-bti.s:
```diff
@@ -0,0 +1,513 @@
+
+/// functions do not have BTI compatible landing pads.
+/// Expect linker generated landing pads.
+.section .text.7, "ax", %progbits
```
We can add a large alignment for `fn1`. `bl fn2` can probably be changed to `b .text.7+offset` to test thunks with a non-zero addend. (https://reviews.llvm.org/D70637)

smithp35 replied:
Added a .text+offset case.

I've not added in the alignment here. I reordered fn1 and fn2 so that I could show that the fallthrough didn't happen with alignment, only to discover that the heuristic  I was using for short thunks wasn't good enough. I've made a change to the heuristic, which I now need a test for.

MaskRay commented on lld/ELF/Arch/AArch64.cpp:
```diff
@@ -28,6 +28,38 @@ uint64_t elf::getAArch64Page(uint64_t expr) {
+    return true;
+  Defined *d = dyn_cast_or_null<Defined>(&s);
+  if (d == nullptr || d->section == nullptr ||
+      d->section->kind() != InputSectionBase::Regular)
```
`!isa_and_nonnull<InputSection>(d->section)`

smithp35 replied:
ACK

MaskRay commented on lld/ELF/Arch/AArch64.cpp:
```diff
@@ -28,6 +28,38 @@ uint64_t elf::getAArch64Page(uint64_t expr) {
+  // PLT entries accessed indirectly have a BTI c.
+  if (s.isInPlt())
+    return true;
+  Defined *d = dyn_cast_or_null<Defined>(&s);
```
`dyn_cast` since `&s` cannot be nullptr.

smithp35 replied:
ACK

MaskRay commented on lld/ELF/Arch/AArch64.cpp:
```diff
@@ -28,6 +28,38 @@ uint64_t elf::getAArch64Page(uint64_t expr) {
+    // the target a BTI landing pad.
+    return true;
+  InputSection *isec = cast<InputSection>(d->section);
+  int64_t off = d->value + a;
```
The sum has a type of `uint64_t`. Using `uint64_t` can avoid `off < 0` below.

smithp35 replied:
ACK

MaskRay commented on lld/ELF/Arch/AArch64.cpp:
```diff
@@ -28,6 +28,38 @@ uint64_t elf::getAArch64Page(uint64_t expr) {
+    return true;
+  const uint8_t *buf = isec->content().begin();
+  const uint32_t instr = read32le(buf + off);
+  if (instr == 0xd503233f || // PACIASP.
```
`return instr == ... || ...`

or perhaps `return is_contained(/*PACIASP*/0xd503233f, ...)`

smithp35 replied:
ACK. I've also used the common parts of the HINT encoding to avoid testing all combinations.

MaskRay commented on lld/ELF/Relocations.cpp:
```diff
@@ -2243,6 +2243,14 @@ std::pair<Thunk *, bool> ThunkCreator::getThunk(InputSection *isec,
 
+std::pair<Thunk *, bool> ThunkCreator::getSyntheticLandingPad(Defined &d,
+                                                              int64_t a) {
+  Thunk *lp = landingPadsBySectionAndAddend[{{d.section, d.value}, a}];
```
Prefer `try_emplace` over `operator[]`

smithp35 replied:
ACK

MaskRay commented on lld/ELF/Relocations.cpp:
```diff
@@ -2326,6 +2334,21 @@ bool ThunkCreator::createThunks(uint32_t pass,
+                auto &dr = cast<Defined>(t->destination);
+                std::tie(lpt, isNew) = getSyntheticLandingPad(dr, t->addend);
+                if (isNew) {
+                  InputSection *targetsec = dyn_cast<InputSection>(dr.section);
```
It seems that `targetSec` could be nullptr in erroneous non-InputSection cases (MergeInputSection) and would cause a null pointer dereference in `getISThunkSec`.

If we don't write code guarding against such cases, perhaps just use `cast<>`.

smithp35 replied:
I've used cast<>

The code in isAArch64BTILandingPad has a test
```
if (!isa_and_nonnull<InputSection>(d->section))
```
so `t->needsSyntheticLandingPad()` should return false in those cases.

MaskRay commented on lld/ELF/Thunks.cpp:
```diff
@@ -578,6 +622,48 @@ void AArch64ADRPThunk::addSymbols(ThunkSection &isec) {
+bool AArch64BTILandingPadThunk::getMayUseShortThunk() {
+  if (!mayUseShortThunk)
+    return false;
+  // If the target is the following instruction then
```
The wrap column is smaller than 80.

smithp35 replied:
ACK. I've reformatted.

MaskRay commented on lld/ELF/Thunks.cpp:
```diff
@@ -578,6 +622,48 @@ void AArch64ADRPThunk::addSymbols(ThunkSection &isec) {
+  addSymbol("$x", STT_NOTYPE, 0, isec);
+}
+
+InputSection *AArch64BTILandingPadThunk::getTargetInputSection() const {
```
This function is not needed.

Relocations.cpp calls `ts = getISThunkSec(targetsec)` without checking `getTargetInputSection`.

smithp35 replied:
It is called in `mergeCmp()` but I've removed it anyway as it is not a strict requirement that the landing pads are before the target section, they just need to be within direct branch range.

smithp35 commented:
Thanks for the review. I'll work on an updated patch, hopefully this week.

smithp35 committed:
Merge branch 'main' into btithunk

smithp35 committed:
Merge branch 'llvm:main' into btithunk

smithp35 committed:
[NFC] Fix incomplete merge.

smithp35 committed:
Address review comments

* Remove getTargetInputSection.
* Simplify AArch64 BTI detection.
* Make short thunk heuristic more robust when there is more than one landing
  pad thunk.

smithp35 committed:
Rework test case

Use more Output sections to avoid needing close to 128 MiB of zeroes.
Permits fewer llvm-objcopy instances as we don't need --start-address etc.

smithp35 commented on lld/test/ELF/aarch64-thunk-bti.s:
```diff
@@ -0,0 +1,513 @@
+// RUN: llvm-mc -filetype=obj -triple=aarch64 %t/shared -o %tshared.o
+// RUN: ld.lld --shared -o %tshared.so %tshared.o
+// RUN: ld.lld %tshared.so --script=%t/lds %t.o -o %t.exe --defsym absolute=0xf0000000
+// RUN: llvm-objdump -d --no-show-raw-insn --print-imm-hex --start-address=0x10000000 --stop-address=0x10000028 %t.exe | FileCheck --check-prefix=CHECK-EXE %s
```
I've had a go at sharing more parts.

smithp35 commented on lld/test/ELF/aarch64-thunk-bti.s:
```diff
@@ -0,0 +1,513 @@
+// place a pool that can short thunk
+// to targets.
+.section .text.1, "ax", %progbits
+.space 0x2000000
```
Thanks for the suggestion, I've used OutputSections and the output is much smaller now.

smithp35 commented on lld/test/ELF/aarch64-thunk-bti.s:
```diff
@@ -0,0 +1,513 @@
+/// alignment requirement. Check that we don't fall through into
+/// alignment padding.
+.section .text.9, "ax", %progbits
+.balign 16
```
In the test I've made sure that we don't end up on an alignment boundary for this case.

smithp35 commented on lld/test/ELF/aarch64-thunk-bti.s:
```diff
@@ -0,0 +1,513 @@
+// CHECK-BTI-NEXT:           ret
+
+/// PLT has bti c.
+// CHECK-SO3: 180000b0: bti     c
```
That's PLT[0] which isn't particularly useful so I've taken it out.

smithp35 commented on lld/test/ELF/aarch64-thunk-bti.s:
```diff
@@ -0,0 +1,513 @@
+// CHECK-EXE3-NEXT:      nop
+
+.section .long_calls, "ax", %progbits
+.global long_calls
```
ACK. Added label.

smithp35 commented on lld/test/ELF/aarch64-thunk-bti.s:
```diff
@@ -0,0 +1,513 @@
+.section .long_calls, "ax", %progbits
+.global long_calls
+.type long_calls, %function
+/// Expect thunk to target as targets have
```
ACK. Reformatted all comments in file.

smithp35 commented:
Thanks for your patience. I've reworked the test case to use a lot fewer zeroes.

MaskRay commented on lld/ELF/Relocations.h:
```diff
@@ -200,6 +203,14 @@ class ThunkCreator {
   // The Mips LA25 Thunk is an example of an inline ThunkSection.
```
Perhaps update this comment to mention `AArch64BTILandingPadThunk`

MaskRay commented on lld/test/ELF/aarch64-thunk-bti.s:
```diff
@@ -0,0 +1,482 @@
+// REQUIRES: aarch64
+// RUN: rm -rf %t && split-file %s %t && cd %t
+// RUN: llvm-mc -filetype=obj -triple=aarch64 asm -o %t.o
```
After `rm -rf %t && split-file %s %t && cd %t`, `%t` can be completely avoided.
For example, name an output file `a.o` and output as `out` instead `%t.exe`

MaskRay commented:
Sorry that Relocations.cpp Thunks.cpp need rebasing due to my `Ctx &ctx` changes.

smithp35 committed:
Merge remote-tracking branch 'upstream/main' into btithunk

smithp35 committed:
Final review comments

Remove spurious semi-colon.
Use structured bindings.
Inline function call to remove variable.
Improve comment about ThunkedSections to include landing pads.
Remove uses of %t in test.

smithp35 commented:
> Sorry that Relocations.cpp Thunks.cpp need rebasing due to my `Ctx &ctx` changes.

Thanks for the warning. I've applied the final set of review comments and rebased. I'll wait till tomorrow before merging.

smithp35 committed:
Fix clang-format differences [NFC]

MaskRay commented on lld/ELF/Thunks.cpp:
```diff
@@ -553,7 +588,11 @@ void AArch64ABSLongThunk::writeLong(uint8_t *buf) {
     0x00, 0x00, 0x00, 0x00,
   };
-  uint64_t s = getAArch64ThunkDestVA(destination, addend);
+  // if mayNeedLandingPad is true then destination is an
```
`// If`

MaskRay commented on lld/ELF/Thunks.cpp:
```diff
@@ -577,7 +616,11 @@ void AArch64ADRPThunk::writeLong(uint8_t *buf) {
       0x00, 0x02, 0x1f, 0xd6, // br   x16
   };
-  uint64_t s = getAArch64ThunkDestVA(destination, addend);
+  // if mayNeedLandingPad is true then destination is an
```
`// If`

MaskRay commented on lld/test/ELF/aarch64-thunk-bti.s:
```diff
@@ -0,0 +1,482 @@
+// RUN: llvm-objdump -d --no-show-raw-insn out.so | FileCheck %s
+// RUN: llvm-objdump -d --no-show-raw-insn out.so | FileCheck %s --check-prefix=CHECK-PADS
+// RUN: llvm-mc -filetype=obj -triple=aarch64 shared -o shared.o
+// RUN: ld.lld --shared -o shared.so shared.o --soname=shared.so
```
`--soname=` could be omitted as we are using relative paths

smithp35 committed:
Further review comments

use {} rather than std::make_pair.
reflow comments.
remove --soname from test.

zmodem commented:
We're hitting asserts after this. I've attached an lld reproducer tarball at https://crbug.com/377438309#comment8

```
$ tar zxf repro.tar.gz
$ ( cd build/repro && ../bin/ld.lld @response.txt )
ld.lld: /work/llvm-project/lld/ELF/Thunks.cpp:621: virtual void (anonymous namespace)::AArch64ADRPThunk::writeLong(uint8_t *): Assertion `!mayNeedLandingPad || landingPad != nullptr' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
 #0 0x000055666b0b8c68 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (../bin/ld.lld+0xc8ac68)
 #1 0x000055666b0b681e llvm::sys::RunSignalHandlers() (../bin/ld.lld+0xc8881e)
 #2 0x000055666b0b9638 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f0f642591a0 (/lib/x86_64-linux-gnu/libc.so.6+0x3d1a0)
 #4 0x00007f0f642a70ec __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007f0f64259102 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #6 0x00007f0f642424f2 abort ./stdlib/abort.c:81:7
 #7 0x00007f0f64242415 _nl_load_domain ./intl/loadmsgcat.c:1177:9
 #8 0x00007f0f64251d32 (/lib/x86_64-linux-gnu/libc.so.6+0x35d32)
 #9 0x000055666b337831 (anonymous namespace)::AArch64ADRPThunk::writeLong(unsigned char*) Thunks.cpp:0:0
#10 0x000055666b2dbf21 lld::elf::ThunkSection::writeTo(unsigned char*) (../bin/ld.lld+0xeadf21)
#11 0x000055666b2a152b void lld::elf::OutputSection::writeTo<llvm::object::ELFType<(llvm::endianness)1, true>>(unsigned char*, llvm::parallel::TaskGroup&)::'lambda'(unsigned long, unsigned long)::operator()(unsigned long, unsigned long) const (../bin/ld.lld+0xe7352b)
#12 0x000055666ef4c785 std::_Function_handler<void (), llvm::parallel::TaskGroup::spawn(std::function<void ()>)::$_0>::_M_invoke(std::_Any_data const&) Parallel.cpp:0:0
#13 0x000055666ef4c263 llvm::parallel::detail::(anonymous namespace)::ThreadPoolExecutor::work(llvm::ThreadPoolStrategy, unsigned int) Parallel.cpp:0:0
#14 0x00007f0f644dee64 (/lib/x86_64-linux-gnu/libstdc++.so.6+0xdee64)
#15 0x00007f0f642a539c start_thread ./nptl/pthread_create.c:444:8
#16 0x00007f0f64326608 clone3 ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:80:0
Aborted
```

Could you take a look?

smithp35 commented:
Thanks for the reproducer. I've downloaded it and reproduced. I'll take a look today.

smithp35 commented:
To give you a quick update on the cause. There is an optimisation that won't create a BTI landing pad for a location if the linker can use a direct branch in the Thunk. https://github.com/llvm/llvm-project/blob/main/lld/ELF/Thunks.cpp#L580 

It looks like what is happening is:
* Pass N thunk is created. It is in short-branch range so a direct branch can be used, no BTI needs creating.
* Pass M (where M > N). More content changes are made so that the thunk can no longer use a short-branch.
* writeTo() I need an indirect branch, where's my landing pad, assert failure.

I need to make a smaller test case, and to make the landing pad generation run each pass. Hopefully I can get this done Today, but just in case it can't, how urgent is this for you?

Right now the hwasan checks
```
0000000000000000 <__hwasan_check_x0_0>:
       0: 9344dc10      sbfx    x16, x0, #4, #52
       4: 38706930      ldrb    w16, [x9, x16]
       8: eb40e21f      cmp     x16, x0, lsr #56
       c: 54000041      b.ne    0x14 <__hwasan_check_x0_0+0x14>
      10: d65f03c0      ret
      14: a9b007e0      stp     x0, x1, [sp, #-0x100]!
      18: a90efbfd      stp     x29, x30, [sp, #0xe8]
      1c: d2800001      mov     x1, #0x0                // =0
      20: 90000010      adrp    x16, 0x0 <__hwasan_check_x0_0>
      24: f9400210      ldr     x16, [x16]
      28: d61f0200      br      x16
```
Do not have a BTI or compatible instruction, so these would cause a runtime failure if BTI was enabled and the linker did generate an indirect branch (and code-path went through it).

There's a quick fix to disable the optimisation, but that would mean updating a rather large test, so I'd prefer to take the time to fix with the optimisation.

smithp35 commented:
I've created https://github.com/llvm/llvm-project/pull/116402 to fix the repro in https://crbug.com/377438309#comment8 I've not got a regression test yet as that will take some time to get one small enough. I hope to add one next week.

zmodem commented:
Thanks for the quick fix!
